src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java

```diff
@@ -310,32 +328,53 @@ void refreshTopologyUnlocked(final Partition[] assignedPartitions) {
                 .filter(p -> p.getState() == Partition.State.REASSIGNING)
                 .collect(Collectors.toMap(Partition::getKey, p -> p));
 
+        boolean modified = false;
+
         // 1. Select which partitions must be removed right now for some (strange) reasons
         // (no need to release topology).
-        offsets.keySet().stream()
+        final List<EventTypePartition> forciblyRemoved = offsets.keySet().stream()
                 .filter(e -> !newAssigned.containsKey(e))
                 .filter(e -> !newReassigning.containsKey(e))
-                .collect(Collectors.toList()).forEach(this::removeFromStreaming);
+                .collect(Collectors.toList());
+        if (!forciblyRemoved.isEmpty()) {
+            modified = true;
+            forciblyRemoved.forEach(this::removeFromStreaming);
```

Comment:
As I understood the state of this partitions UNASSIGNED, why do we want to remove them from streaming ?

Reply:
Suppose that subscription was streaming p0 and p1. 
Then, for some reason, refresh comes and says that now it's time to stream p0 only.
The only thing to do there is to react on this change. That is what is going on - in case when partition is not ours anymore - remove it from streaming.

Reply:
Okay, could you fill the 3 arrays in 1 loop not in 3?